drivers/crypto/crypto_smartbond.c

```diff
@@ -0,0 +1,1049 @@
+/*
+ * Copyright (c) 2023 Renesas Electronics Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/irq.h>
+#include <DA1469xAB.h>
+#include <da1469x_config.h>
+#include <da1469x_otp.h>
+#include <system_DA1469x.h>
+
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(crypto_smartbond_crypto);
+
+#define DT_DRV_COMPAT renesas_smartbond_crypto
+
+#define IRQN      DT_INST_IRQN(0)
+#define IRQ_PRIO  DT_INST_IRQ(0, priority)
+
+#if defined(CONFIG_CRYPTO_ASYNC)
+#define CRYPTO_HW_CAPS  (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_ASYNC_OPS | CAP_NO_IV_PREFIX)
+#else
+#define CRYPTO_HW_CAPS  (CAP_RAW_KEY | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS | CAP_NO_IV_PREFIX)
+#endif
+
+#define CRYPTO_MAX_SESSION  CONFIG_CRYPTO_SMARTBOND_MAX_SESSION
+
+#define SWAP32(_w)   __builtin_bswap32(_w)
+
+/* Forward declarations to suppress compiler warnings */
+struct crypto_smartbond_data;
+struct crypto_smartbond_session;
+
+
+#define CRYPTO_CTRL_REG_SET(_field, _val) \
+	AES_HASH->CRYPTO_CTRL_REG = \
+	(AES_HASH->CRYPTO_CTRL_REG & ~AES_HASH_CRYPTO_CTRL_REG_ ## _field ## _Msk) | \
+	((_val) << AES_HASH_CRYPTO_CTRL_REG_ ## _field ## _Pos)
+
+#define CRYPTO_CTRL_REG_GET(_field) \
+	((AES_HASH->CRYPTO_CTRL_REG & AES_HASH_CRYPTO_CTRL_REG_ ## _field ## _Msk) >> \
+	AES_HASH_CRYPTO_CTRL_REG_ ## _field ## _Pos)
+
+
+struct crypto_smartbond_data {
+    /*
+     * Semaphore to provide mutual exlusion when a crypto session is requested.
+     */
+	struct k_sem session_sem;
+
+    /*
+     * Semaphore to provide mutual exlusion when a cryptographic task is requested.
+     * (a session should be requested at this point).
+     */
+	struct k_sem device_sem;
+#if defined(CONFIG_CRYPTO_ASYNC)
+    /*
+     * User-defined callbacks to be called upon completion of asynchronous
+     * cryptographic operations. Note that the AES and HASH modes can work
+     * complementary to each other.
+     */
+	union {
+		cipher_completion_cb cipher_user_cb;
+		hash_completion_cb hash_user_cb;
+	};
+
+    /*
+     * Packet context should be stored during a session so that can be rertieved
+     * from within the crypto engine ISR context.
+     */
+	union {
+		struct cipher_pkt *cipher_pkt;
+		struct hash_pkt *hash_pkt;
+	};
+#else
+    /*
+     * Semaphore used to block for as long as a synchronous cryptographic operation
+     * is in progress.
+     */
+	struct k_sem sync_sem;
+#endif
+};
+
+struct crypto_smartbond_session {
+    /*
+     * Status flag to indicate if the crypto engine resources have been granted. Note that the
+     * device integrates a single crypto engine instance.
+     */
+	bool in_use;
+};
+
+static struct crypto_smartbond_session crypto_smartbond_sessions[CRYPTO_MAX_SESSION];
```

Comment:
I chose that approach just in case another family of device with the same Crypto block integrates multiple instances. Other drivers also follow that approach. In any case, CRYPTO_MAX_SESSION should now be resolved to 1.

Reply:
it's not clear how multiple sessions would be handled by new hw and since the session struct doesn't store any session data at the moment, it's redundant. it's enough to just have global `is_use` flag like mentioned `nrf_ecb` driver has. the support for multiple sessions can be added properly once hw is available.

Reply:
Done